starboard/nplb/maximum_player_configuration_explorer.cc

```diff
@@ -0,0 +1,242 @@
+      << " Exploring this solution space may be a time-consuming process.";
+  SB_DCHECK(test_functor);
+
+  Run();
+}
+
+void GradedPosetDFSScheduler::Run() {
+  std::deque<std::pair<std::vector<int>, int>> queue;
+  std::unordered_map<std::vector<int>, bool, HashFunction> advancable;
+
+  queue.push_front(std::make_pair(std::vector<int>(resources_, 0), 0));
+  advancable[queue.front().first] = false;
+  while (!queue.empty()) {
+    auto& it = queue.front();
+    auto& config = it.first;
+    auto& index = it.second;
+    while (index < resources_) {
+      if (config[index] != instances_) {
+        ++config[index];
+        if (advancable.find(config) != advancable.end()) {
+          --config[index];
+        } else {
+          break;
+        }
+      }
+      ++index;
+    }
+
+    if (index != resources_) {
+      if (test_functor_(config)) {
+        queue.push_front(std::make_pair(config, index));
+        advancable[config] = true;
+      } else {
+        advancable[config] = false;
+      }
+      --config[index];
+      ++index;
+    } else {
+      bool can_advance = false;
+      for (int i = 0; i < config.size(); ++i) {
+        if (config[i] >= instances_) {
+          continue;
+        } else {
+          ++config[i];
+          if (advancable[config]) {
+            can_advance = true;
+            break;
+          }
+          --config[i];
+        }
+      }
+      if (!can_advance) {
+        max_test_config_set_.insert(config);
+      }
+      queue.pop_front();
+    }
+  }
+}
+
+MaximumPlayerConfigurationExplorer::MaximumPlayerConfigurationExplorer(
+    const std::set<PlayerConfig>& player_configs,
+    testing::FakeGraphicsContextProvider* fake_graphics_context_provider)
+    : fake_graphics_context_provider_(fake_graphics_context_provider) {
+  SB_DCHECK(!player_configs.empty());
+
+  player_configs_.assign(player_configs.begin(), player_configs.end());
+  Init();
+}
+
+MaximumPlayerConfigurationExplorer::MaximumPlayerConfigurationExplorer(
+    const std::set<PlayerConfig>& player_configs,
+    const std::set<std::vector<int>>& pseudo_max_element_set,
+    testing::FakeGraphicsContextProvider* fake_graphics_context_provider)
+    : fake_graphics_context_provider_(fake_graphics_context_provider) {
+  SB_DCHECK(!player_configs.empty());
+  SB_DCHECK(!pseudo_max_element_set.empty());
+
+  player_configs_.assign(player_configs.begin(), player_configs.end());
+  pseudo_max_element_set_.assign(pseudo_max_element_set.begin(),
+                                 pseudo_max_element_set.end());
+  for (int i = 0; i < pseudo_max_element_set_.size(); ++i) {
+    for (int j = i + 1; j < pseudo_max_element_set_.size(); ++j) {
+      SB_DCHECK(!PosetInSameChain(pseudo_max_element_set_[i],
+                                  pseudo_max_element_set_[j]));
+    }
+  }
+
+  Init();
+}
+
+MaximumPlayerConfigurationExplorer::~MaximumPlayerConfigurationExplorer() {
+  for (const auto& players_of_config : players_) {
```

Comment:
Consider using

```
  // |players_| is a two dimensional vector of SbPlayer instances.
  for (int i = 0; i < players_.size(); ++i) {
    for (auto&& player : players[i]) {
      SbPlayerDestroy(player);
    }
  }
```

This can avoid the naming of `players_of_config`.

Reply:
done